core/anvil/src/region/mod.rs

```diff
@@ -31,62 +31,232 @@ const DATA_VERSION: i32 = 1631;
+    pub(crate) fn pack_u9(u9_array: Vec<u16>) -> Vec<i64> {
+        let mut packed = Vec::with_capacity(36);
+        let mut iter = u9_array.iter();
+
+        let mut container = 0u64;
+        let mut shift = 0;
+        for _elem in 0..256 {
+            // For every element (u9)
+            let element = *iter.next().unwrap() as u64;
+
+            if element > 0x1FF {
+                // Invalid heightmap value
+                panic!("Invalid heightmap value {}", element);
+            }
+
+            container |= element << shift;
+            shift += 9;
+
+            if shift >= 64 {
+                // Take next container
+                packed.push(container as i64);
+
+                container = if shift > 64 {
+                    // We have some bits left to store in the new container
+                    element >> -(shift - 64 - 9)
+                } else {
+                    0
+                };
+
+                shift -= 64;
+            }
+        }
+
+        debug_assert_eq!(iter.next(), None);
+        debug_assert_eq!(shift, 0);
+        debug_assert_eq!(container, 0);
+
+        packed
+    }
+}
+
+mod packed_u9 {
+    use serde::de::Error as DeError;
+    use serde::de::{SeqAccess, Visitor};
+    use serde::export::Formatter;
+    use serde::ser::Error as SerError;
+    use serde::ser::SerializeSeq;
+    use serde::{Deserializer, Serializer};
+    use std::marker::PhantomData;
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<u16>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct PackedVisitor(PhantomData<fn() -> u16>);
+
+        impl<'de> Visitor<'de> for PackedVisitor {
+            type Value = Vec<u16>;
+
+            fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+                formatter.write_str("a sequence of type long with length 36")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let len = seq.size_hint().unwrap(); // nbt always knows sequence size
+                if len != 36 {
+                    // Invalid sequence length
+                    return Err(A::Error::custom("sequence length not equal to 36"));
+                }
+
+                let mut u9_array: Vec<u16> = Vec::with_capacity(256);
+
+                let mut container: Option<u64> = seq.next_element()?.map(|x: i64| x as u64); // We checked the length
+                let mut shift = 0;
+                for _elem in 0..256 {
+                    // For every element (u9)
+                    let mut element: u16 = ((container.unwrap() >> shift) & 0x1FF) as u16;
+                    shift += 9;
+
+                    if shift >= 64 {
+                        // Take next container
+                        container = seq.next_element()?.map(|x: i64| x as u64);
+
+                        element |= if shift > 64 {
```

Comment:
I think

```rust
if shift > 64 {
    element |= container.unwrap() << -(shift - 64 - 9) & 0x1FF;
}
```
is equivalent here, and the intent is more clear.

(Rather than bitwise ORing with 0 if the condition is false.)

Reply:
Interesting that clippy doesn't catch stuff like that